feat(set_theory): add to cardinal, ordinal, cofinality

fpvandoorn opened the pull request:
The main new fact is the infinite pigeonhole principle
Also includes many basic additions

TO CONTRIBUTORS:

Make sure you have:

  * [x] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/naming.md)
  * [x] make sure definitions and lemmas are put in the right files
  * [x] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/code-review.md)

fpvandoorn committed:
feat(set_theory): add to cardinal, ordinal, cofinality

The main new fact is the infinite pigeonhole principle
Also includes many basic additions

fpvandoorn committed:
fix name change in other files

digama0 commented on src/algebra/order.lean:
```diff
@@ -35,6 +35,9 @@ lt_of_le_not_le ((le_total _ _).resolve_right h) h
 ⟨not_le_of_gt, lt_of_not_ge'⟩
 
+lemma le_iff_not_gt [linear_order α] (x y : α) : x ≤ y ↔ ¬x > y :=
+⟨not_lt_of_ge, le_of_not_gt⟩
```
What is this used for? It should be superceded by `not_le`

fpvandoorn replied:
I removed it

digama0 commented on src/order/basic.lean:
```diff
@@ -449,27 +449,70 @@ def unbounded (r : α → α → Prop) (s : set α) : Prop := ∀ a, ∃ b ∈ s
+  simp only [bounded, unbounded, not_forall, not_exists, exists_prop, not_and, not_not]
+end
+
+@[simp] lemma not_unbounded_iff {r : α → α → Prop} (s : set α) : ¬unbounded r s ↔ bounded r s :=
```
this is `not_iff_comm` applied to the previous theorem. Also you can use `classical` in both proofs rather than `haveI := classical.prop_decidable`

fpvandoorn replied:
What do you mean by "use `classical`"?

fpvandoorn replied:
Do you mean `local attribute [instance, priority 0] classical.prop_decidable`?

digama0 replied:
I mean the `classical` tactic

digama0 commented on src/order/basic.lean:
```diff
@@ -449,27 +449,70 @@ def unbounded (r : α → α → Prop) (s : set α) : Prop := ∀ a, ∃ b ∈ s
+min_mem wf { x | ∀a ∈ s, r a x } (ne_empty_iff_exists_mem.mpr h) x hx
+
+protected noncomputable def succ {α} {r : α → α → Prop} (wf : well_founded r) (x : α)
+  (h : ∃y, r x y) : α :=
```
Can `h` be removed from the definition? If there is no such y, then you can define `succ x = x`

digama0 commented on src/set_theory/cardinal.lean:
```diff
@@ -53,6 +54,9 @@ theorem le_mk_iff_exists_set {c : cardinal} {α : Type u} :
   ⟨set.range f, eq.symm $ quot.sound ⟨equiv.set.range f hf⟩⟩,
 λ ⟨p, e⟩, e ▸ ⟨⟨subtype.val, λ a b, subtype.eq⟩⟩⟩
 
+def out_embedding {c c' : cardinal} : c ≤ c' ↔ nonempty (c.out ↪ c'.out) :=
```
this is a `theorem`

digama0 commented on src/set_theory/cardinal.lean:
```diff
@@ -622,6 +632,22 @@ match a, b, lt_omega.1 ha, lt_omega.1 hb with
+  rintro ⟨f'⟩, cases embedding.trans f' equiv.ulift.to_embedding with f hf, exact ⟨f, hf⟩
+end
+
+lemma exists_two_elements (h : (2 : cardinal) ≤ mk α) : ∃x y : α, x ≠ y :=
```
This is an iff, and it should be `two_le_iff`

digama0 commented on src/set_theory/cardinal.lean:
```diff
@@ -622,6 +632,22 @@ match a, b, lt_omega.1 ha, lt_omega.1 hb with
+lemma exists_two_elements (h : (2 : cardinal) ≤ mk α) : ∃x y : α, x ≠ y :=
+by { cases h with f, refine ⟨f $ sum.inl ⟨⟩, f $ sum.inr ⟨⟩, _⟩, intro h, cases f.2 h }
+
+lemma exists_unequal_element (h : (2 : cardinal) ≤ mk α) (x : α) : ∃y : α, x ≠ y :=
```
This is also an iff. I don't have any other ideas for the name, maybe just `two_le_iff'`

digama0 commented on src/set_theory/cardinal.lean:
```diff
@@ -725,4 +769,126 @@ quot.sound ⟨equiv.set.union (disjoint_iff.1 H)⟩
 lemma mk_le_mk_of_subset {α} {s t : set α} (h : s ⊆ t) : mk s ≤ mk t :=
 ⟨ set.embedding_of_subset h ⟩
 
+lemma mk_le_of_subproperty {p q : α → Prop} (h : ∀x, p x → q x) : mk {x // p x} ≤ mk {x // q x} :=
```
I suggest `mk_subtype_mono`

digama0 commented on src/set_theory/cardinal.lean:
```diff
@@ -725,4 +769,126 @@ quot.sound ⟨equiv.set.union (disjoint_iff.1 H)⟩
+⟨⟨subtype.val, subtype.val_injective⟩⟩
+
+lemma mk_image_eq_lift {α : Type u} {β : Type v} (f : α → β) (s : set α) (h : injective f) :
+  lift.{v u} (mk (f '' s)) = lift.{u v} (mk s) :=
```
You should be able to use `lift_mk_eq` rather than dealing with `ulift`

digama0 commented on src/set_theory/cardinal.lean:
```diff
@@ -725,4 +769,126 @@ quot.sound ⟨equiv.set.union (disjoint_iff.1 H)⟩
+quotient.sound ⟨equiv.ulift.trans ((equiv.set.image f s h).symm.trans equiv.ulift.symm)⟩
+
+lemma mk_image_eq_of_inj_on {α β : Type u} (f : α → β) (s : set α) (h : inj_on f s) :
+  mk (f '' s) = mk s :=
```
This should follow similarly to the previous theorem; possibly a version of `equiv.set.image` using `inj_on` is needed

digama0 commented on src/set_theory/cardinal.lean:
```diff
@@ -725,4 +769,126 @@ quot.sound ⟨equiv.set.union (disjoint_iff.1 H)⟩
+quotient.sound ⟨equiv.subtype_equiv_of_subtype' e⟩
+
+lemma mk_sep (s : set α) (t : α → Prop) : mk ({ x ∈ s | t x } : set α)  = mk { x : s | t x.1 } :=
+by { refine quotient.sound ⟨_⟩, symmetry, apply (equiv.subtype_subtype_equiv_subtype _ _).trans _,
```
Why isn't the proof just `quotient.sound ⟨equiv.subtype_subtype_equiv_subtype _ _⟩`? Possibly another equiv theorem is needed

fpvandoorn replied:
We have to rewrite along `exists_prop : (∃ h : p, q) ↔ p ∧ q`. I'll add another theorem in `equiv.basic`

digama0 commented on src/set_theory/cardinal.lean:
```diff
@@ -725,4 +769,126 @@ quot.sound ⟨equiv.set.union (disjoint_iff.1 H)⟩
+     simp only [exists_prop, mem_sep_eq, set_coe_eq_subtype, mem_set_of_eq] }
+
+lemma mk_preimage_of_injective_lift {α : Type u} {β : Type v} (f : α → β) (s : set β)
+  (h : injective f) : lift.{u v} (mk (f ⁻¹' s)) ≤ lift.{v u} (mk s) :=
```
Use `lift_mk_le` here. also in the next theorem

digama0 commented on src/set_theory/cardinal.lean:
```diff
@@ -725,4 +769,126 @@ quot.sound ⟨equiv.set.union (disjoint_iff.1 H)⟩
+end
+
+lemma mk_preimage_of_onto_lift {α : Type u} {β : Type v} (f : α → β) (s : set β)
+  (h : s ⊆ range f) : lift.{v u} (mk s) ≤ lift.{u v} (mk (f ⁻¹' s)) :=
```
Not sure about the use of `onto`, we've never called it that before. Maybe `surj` instead? Or just `subset_range`

digama0 commented on src/set_theory/cardinal.lean:
```diff
@@ -725,4 +769,126 @@ quot.sound ⟨equiv.set.union (disjoint_iff.1 H)⟩
+infix ` ^< `:80 := powerlt
+
+def powerlt_helper {c c' : cardinal} (h : c < c') :
+  ∃(s : {s : set c'.out // mk s < c'}), mk s = c :=
```
This is a `theorem`. Also the conventional term is `aux` not `helper`

digama0 commented on src/set_theory/cardinal.lean:
```diff
@@ -725,4 +769,126 @@ quot.sound ⟨equiv.set.union (disjoint_iff.1 H)⟩
+lemma powerlt_max {c₁ c₂ c₃ : cardinal} : c₁ ^< max c₂ c₃ = max (c₁ ^< c₂) (c₁ ^< c₃) :=
+by { cases le_total c₂ c₃; simp only [max_eq_left, max_eq_right, h, powerlt_le_powerlt_left] }
+
+lemma zero_powerlt {a : cardinal} (h : a ≠ 0) : 0 ^< a = 1 :=
```
You should show what `a ^< 0` is

digama0 commented on src/set_theory/cofinality.lean:
```diff
@@ -51,17 +65,19 @@ namespace ordinal
 end
 
+lemma cof_type (r : α → α → Prop) [is_well_order α r] : (type r).cof = strict_order.cof r :=
+rfl
```
put `rfl` on the same line

digama0 commented on src/set_theory/cofinality.lean:
```diff
@@ -311,6 +327,85 @@ le_antisymm (cof_le_card _) begin
 end
 
+theorem sup_lt_ord {ι} (f : ι → ordinal) {c : cardinal} (H1 : cardinal.mk ι < c.ord.cof)
+  (H2 : ∀ i, f i < c.ord) : sup.{u u} f < c.ord :=
```
Does this theorem need `ord`? I think it's true for any ordinal

digama0 commented on src/set_theory/ordinal.lean:
```diff
@@ -481,6 +499,9 @@ theorem type_eq {α β} {r : α → α → Prop} {s : β → β → Prop}
   [is_well_order α r] [is_well_order β s] :
   type r = type s ↔ nonempty (r ≃o s) := quotient.eq
 
+lemma type_out (o : ordinal) : type o.out.r = o :=
```
This can be a simp lemma

digama0 commented on src/set_theory/ordinal.lean:
```diff
@@ -574,13 +617,17 @@ theorem typein_surj (r : α → α → Prop) [is_well_order α r]
 induction_on o (λ β s _ ⟨f⟩, by exactI ⟨f.top, typein_top _⟩) h
 
+lemma injective_typein (r : α → α → Prop) [is_well_order α r] : injective (typein r) :=
+injective_of_increasing r (<) (typein r) (λ x y, (typein_lt_typein r).2)
```
Either this should be proven from `typein_inj`, or `typein_inj` should be shortened using this

digama0 commented on src/set_theory/ordinal.lean:
```diff
@@ -636,6 +704,10 @@ quot.lift_on o (λ ⟨α, r, _⟩, mk α) $
 
+lemma card_typein {r : α → α → Prop} [wo : is_well_order α r] (x : α) :
+  mk {y // r y x} = (typein r x).card :=
+rfl
```
rfl on same line

digama0 commented on src/set_theory/ordinal.lean:
```diff
@@ -963,6 +1031,14 @@ instance : decidable_linear_order ordinal :=
 
+@[simp] lemma typein_le_typein (r : α → α → Prop) [is_well_order α r] {x x' : α} :
+  typein r x ≤ typein r x' ↔ ¬r x' x :=
+by rw [le_iff_not_gt, not_iff_not, gt, typein_lt_typein]
```
if you use `<- not_lt` here you don't need to unfold `gt`. Also `not_iff_not` is not necessary

digama0 commented on src/set_theory/ordinal.lean:
```diff
@@ -1140,6 +1216,27 @@ end
        dif_neg h.1, dif_neg (not_succ_of_is_limit h)]; refl
 
+lemma has_succ_of_is_limit {α} {r : α → α → Prop} [wo : is_well_order α r]
+  (h : (type r).is_limit) (x : α) : ∃y, r x y :=
```
What's the motivation for this? I would hope that you don't need to work with the internals of `ordinal` as much as you are

fpvandoorn replied:
Well, sometimes you want to work with an initial segment of ordinals (mathematically), and do things like taking suprema. Because of universe issues you have to live in a well-order which has as order type some ordinal. Now you start with the assumption that the ordinal is a limit ordinal, and now I want to show that I can take a supremum of successors of elements of `\alpha`. This lemma shows that I can take the successor of an element.

digama0 replied:
Can't you use `enum` to achieve this? That is, you might need the well order type but you can still use ordinals to index them. Or is it too cumbersome? I see you added a lemma about `o.out` as a well ordered type; maybe we can just work with this?

fpvandoorn replied:
I now rewrote the proof of this to use `enum`. 
It is still two complicated lines to proof this, so I think this lemma has a place in the library, because it would be too cumbersome to do this every time you need this fact.

fpvandoorn committed:
address all of Mario's comments

fpvandoorn committed:
use classical tactic in order/basic

I did not use it for well_founded.succ, because that resulted in an error in lt_succ

fpvandoorn committed:
fix error

jcommelin committed:
Merge branch 'master' into ordinal_pr

Commit:
Merge branch 'master' into 'ordinal_pr'
